src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -95,19 +120,51 @@ public Http2Connection(Http2ConnectionContext context)
 
         public void OnInputOrOutputCompleted()
         {
-            _stopping = true;
+            lock (_connectionStateLock)
+            {
+                if (_connectionState != ConnectionState.Closed)
+                {
+                    _frameWriter.WriteGoAwayAsync(_highestOpenedStreamId, Http2ErrorCode.NO_ERROR);
+                    _connectionState = ConnectionState.Closed;
+                }
+            }
+
             _frameWriter.Complete();
         }
 
         public void Abort(ConnectionAbortedException ex)
         {
-            _stopping = true;
+            lock (_connectionStateLock)
+            {
+                if (_connectionState != ConnectionState.Closed)
+                {
+                    _frameWriter.WriteGoAwayAsync(_highestOpenedStreamId, Http2ErrorCode.INTERNAL_ERROR);
+                    _connectionState = ConnectionState.Closed;
+                }
+            }
+
             _frameWriter.Abort(ex);
         }
 
         public void StopProcessingNextRequest()
+            => StopProcessingNextRequest(true);
+
+        public void StopProcessingNextRequest(bool sendGoAway = false)
         {
-            _stopping = true;
+            lock (_connectionStateLock)
+            {
+                if (_connectionState == ConnectionState.Open)
+                {
+                    if (sendGoAway)
+                    {
+                        _frameWriter.WriteGoAwayAsync(Int32.MaxValue, Http2ErrorCode.NO_ERROR);
+                    }
+
+                    _connectionState = ConnectionState.Closing;
```

Comment:
So another goaway frame will be written once the last active stream completes, right Could you move directly to the closed state (and maybe write a second goaway frame immediately) in the case there aren't any active streams at this point?

Reply:
Good point.